### PR TITLE
TCP reassembly - support closing connection with FIN + RST on one side

### PR DESCRIPTION
**Pull request by prudens**
Add check logic code to ensure that when the client sends both FIN and RST commands, the server can also immediately close, ensuring that the entire process is in a closed state.

### CODE REVIEW

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 1]
```diff
@@ -1264,3 +1264,36 @@ PTF_TEST_CASE(TestTcpReassemblyTimeStamps)
...
+
+	pcpp::TcpReassemblyConfiguration config(true, 2, 1);
+	pcpp::TcpReassembly tcpReassembly(tcpReassemblyMsgReadyCallback, &results, tcpReassemblyConnectionStartCallback,
+	                                  tcpReassemblyConnectionEndCallback, config);
+
+	std::vector<pcpp::RawPacket> packetStream;
+	PTF_ASSERT_TRUE(
+	    readPcapIntoPacketVec("PcapExamples/one_tcp_stream_fin_rst_close_packet.pcap", packetStream, errMsg));
+
+	pcpp::RawPacket lastPacket = packetStream.back();
```
Comment (tigercosmos) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1281]: unused?

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 2]
```diff
@@ -1264,3 +1264,36 @@ PTF_TEST_CASE(TestTcpReassemblyTimeStamps)
...
+
+	packetStream.pop_back();
```
Comment (tigercosmos) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1283]: why do you pop the last packet?

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 3]
```diff
@@ -1264,3 +1264,36 @@ PTF_TEST_CASE(TestTcpReassemblyTimeStamps)
...
+
+	for (auto iter : packetStream)
```
Comment (tigercosmos) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1285]: let's declare `tcpReassembly` before this line.

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 4]
```diff
@@ -1264,3 +1264,36 @@ PTF_TEST_CASE(TestTcpReassemblyTimeStamps)
...
+	{
+		pcpp::Packet packet(&iter);
+		tcpReassembly.reassemblePacket(packet);
+	}
+
+	pcpp::TcpReassembly::ConnectionInfoList managedConnections =
```
Comment (tigercosmos) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1291]: let's use `auto` to shorten the line.

File: Packet++/src/TcpReassembly.cpp [hunk 5]
```diff
@@ -227,6 +227,11 @@ TcpReassembly::ReassemblyStatus TcpReassembly::reassemblePacket(Packet& tcpData)
 	if (tcpReassemblyData->twoSides[sideIndex].gotFinOrRst)
 	{
 		PCPP_LOG_DEBUG("Got a packet after FIN or RST were already seen on this side (" << static_cast<int>(sideIndex) << "). Ignoring this packet");
+		if (!tcpReassemblyData->twoSides[1 - sideIndex].gotFinOrRst && isRst)
```
Comment (tigercosmos) [inline Packet++/src/TcpReassembly.cpp:230]: not very familiar with this piece of code, could you explain why it's `1 - sideIndex` here?

Comment (prudens) [reply Packet++/src/TcpReassembly.cpp:230]: 1-sizeIndex is mean other side tcp stream，the sideIndex's value maybe 0 or 1.

File: Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp [hunk 6]
```diff
@@ -1264,3 +1264,31 @@ PTF_TEST_CASE(TestTcpReassemblyTimeStamps)
...
+{
+	TcpReassemblyMultipleConnStats results;
+	std::string errMsg;
+
+	std::vector<pcpp::RawPacket> packetStream;
+	PTF_ASSERT_TRUE(readPcapIntoPacketVec("PcapExamples/one_tcp_stream_fin_rst_close_packet.pcap", packetStream, errMsg));
+
+	pcpp::TcpReassemblyConfiguration config(true, 2, 1);
+	pcpp::TcpReassembly tcpReassembly(tcpReassemblyMsgReadyCallback, &results, tcpReassemblyConnectionStartCallback,
+	                                  tcpReassemblyConnectionEndCallback, config);
```
Comment (seladb) [inline Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp:1278]: I think a better approach would be to call `tcpReassemblyTest(packetStream, tcpReassemblyResults, true, false)` and check the results:
- Verify there is only 1 connection
- Verify that `connectionsStarted == true`, `connectionsEnded == false`, `connectionsEndedManually == false`
- Verify that the number of packets from each side is what we expect

**Comment by seladb** [comment]
@prudens can you please address the review comments, and also resolve the conflicts in ` Tests/Pcap++Test/Tests/TcpReassemblyTests.cpp`?

**Comment by seladb** [comment]
@prudens we added clang-format so you need to run it on your code.
Here how to install clang-format: https://github.com/seladb/PcapPlusPlus/blob/0a843116d6f679f6dcc9f9c3c002ca94b32c227f/.github/workflows/build_and_test.yml#L33

```shell
python3 -m pip install cmake-format clang-format==18.1.6
```

**Comment by seladb** [comment]
@prudens can you please address these 2 comment:
- https://github.com/seladb/PcapPlusPlus/pull/1496#discussion_r1683714487
- https://github.com/seladb/PcapPlusPlus/pull/1496#discussion_r1683963061

File: Packet++/src/TcpReassembly.cpp [hunk 7]
```diff
@@ -227,6 +227,11 @@ TcpReassembly::ReassemblyStatus TcpReassembly::reassemblePacket(Packet& tcpData)
 	if (tcpReassemblyData->twoSides[sideIndex].gotFinOrRst)
 	{
 		PCPP_LOG_DEBUG("Got a packet after FIN or RST were already seen on this side (" << static_cast<int>(sideIndex) << "). Ignoring this packet");
```
[the remark this answers is not in the document]
Comment (seladb) [reply Packet++/src/TcpReassembly.cpp]: Fixed in this commit https://github.com/seladb/PcapPlusPlus/pull/1496/commits/a1f3495f0780beb31b7633053f47b1d814dbabe9

**Review by seladb — APPROVED** [review]

**Pull request merged**